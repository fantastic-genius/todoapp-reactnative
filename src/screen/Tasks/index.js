import React, { useEffect, useState, useLayoutEffect } from 'react';
import { 
  View,
  Text,
  StyleSheet,
  SafeAreaView,
  TouchableOpacity,
  FlatList,
  ActivityIndicator,
  Dimensions
} from 'react-native';
import { Entypo, FontAwesome } from '@expo/vector-icons'
import { useDispatch, useSelector } from 'react-redux';

//components
import HeaderTitle from '../../components/HeaderTitle';
import Task from '../../components/Task';

//actions
import { getTasks, getTasksCleanup } from '../../store/actions/getTasks';

const Tasks = ({navigation}) => {
  const dispatch = useDispatch();
  const getTasksState = useSelector(s => s.getTasks);
  const [tasks, setTasks] = useState([]);
  const [selected, setSelected] = useState({});

  const renderTasks = ({item}) => (
    <Task
      task={item}
      key={item.id}
      handleSelect={() => {
        if(!selected.hasOwnProperty(item.id)){
          setSelected(prev => ({...prev, [item.id]: true}))
        }else{
          setSelected(prev => ({...prev, [item.id]: !selected[item.id]}))
        }
      }}
      checked={selected[item.id]}
    />
  )

  useEffect(() => {
    dispatch(getTasksCleanup());
    dispatch(getTasks());
  },[])

  useEffect(() => {
    if(getTasksState.isSuccessful){
      setTasks(getTasksState.data.tasks)
      dispatch(getTasksCleanup());
    }
  },[getTasksState])

  return (
    <SafeAreaView 
      style={styles.container}
    >
      <View style={styles.header}>
        <TouchableOpacity>
          <Entypo name='menu' size={24} color='#d3d7de' />
        </TouchableOpacity>
        <HeaderTitle title='todo' />
        <TouchableOpacity>
          <FontAwesome name='comment-o' size={24} color='#d3d7de' />
        </TouchableOpacity>
      </View>
      <View style={styles.subHeader}>
        <Text style={styles.text}>Tasks</Text>
        <TouchableOpacity style={styles.plusBtn}>
          <Entypo name='plus' size={24} color='white' />
        </TouchableOpacity>
      </View>
      <View style={styles.body}>
        {getTasksState.isLoading ? (
          <ActivityIndicator size='large' color={'#5c677c'} style={styles.loader} />
        ) : (
          <FlatList
            keyExtractor={(item) => item.id}
            data={tasks}
            renderItem={renderTasks}
            ListEmptyComponent={getTasksState.error ? (
              <Text style={styles.text}>
                Something went wrong. Please
                <Text
                  onPress={() => {
                    dispatch(getTasksCleanup());
                    dispatch(getTasks());
                  }}
                  style={styles.retry}
                >
                  {' '}Retry
                </Text>
              </Text>
            ) :( 
              <Text style={styles.text}>No task has been created</Text>
            )}
          />
        )}
      </View>
    </SafeAreaView>
  )
};

const styles = StyleSheet.create({
  container: {
    flex: 1, 
    width: '100%'
  },
  header: {
    backgroundColor: '#3d4452',
    paddingHorizontal: 15,
    height: 50,
    marginTop: 40,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  subHeader: {
    backgroundColor: '#f9fafc',
    paddingHorizontal: 15,
    height: 50,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  text: {
    color: '#717171',
    fontSize: 13
  },
  body: {
    backgroundColor: '#f3f5f9',
    padding: 15
  },
  plusBtn: {
    backgroundColor: '#92cf5c',
    height: 30,
    width: 30,
    borderRadius: 3,
    alignItems: 'center',
    justifyContent: 'center'
  },
  loader: {
    height: Dimensions.get('window').height * 0.3
  },
  retry: {
    color: 'skyblue'
  }
});

export default Tasks;
