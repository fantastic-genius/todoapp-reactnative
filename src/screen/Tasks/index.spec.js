import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux'
import Tasks from './index';
import configureStore from 'redux-mock-store';

const middlewares = []
const mockStore = configureStore(middlewares)
const initialState = {
  getTasks: {
    isLoading: false,
    isSuccessful: true,
    data: {
      "tasks": [
        {"id":"1", "name":"Task 1 ", "due_date":"9:00AM 12-1-2019", "status":"done"},
        {"id":"2", "name":"Task 2 ", "due_date":"11:00AM 12-1-2019", "status":"in-progress"},
        {"id":"3", "name":"Task 3 ", "due_date":"22:00AM 12-1-2019", "status":"pending"}
      ],
      "all_tasks_completed": false,
      "username":"User"
    },
    error: null
  },
};

const store = mockStore(initialState)

describe('<Tasks />', () => {
  it('Has three children', async () => {
    const component = renderer.create(
      <Provider store={store}>
        <Tasks />
      </Provider>
    ).toJSON();
    expect(component).toMatchSnapshot();
  });
});

