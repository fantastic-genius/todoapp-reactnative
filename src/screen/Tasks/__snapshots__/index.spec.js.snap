// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tasks /> Has three children 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#3d4452",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "marginTop": 40,
        "paddingHorizontal": 15,
      }
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#d3d7de",
            "fontSize": 25,
            "lineHeight": 50,
          }
        }
      >
        todo
      </Text>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#f9fafc",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "paddingHorizontal": 15,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#717171",
          "fontSize": 13,
        }
      }
    >
      Tasks
    </Text>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#92cf5c",
          "borderRadius": 3,
          "height": 30,
          "justifyContent": "center",
          "opacity": 1,
          "width": 30,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#f3f5f9",
        "padding": 15,
      }
    }
  >
    <RCTScrollView
      ListEmptyComponent={
        <Text
          style={
            Object {
              "color": "#717171",
              "fontSize": 13,
            }
          }
        >
          No task has been created
        </Text>
      }
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <Text
          onLayout={[Function]}
          style={
            Object {
              "color": "#717171",
              "fontSize": 13,
            }
          }
        >
          No task has been created
        </Text>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
